### PR TITLE
[feat][api] Prebuilt API redesign and simplification

### PR DESCRIPTION
**Pull request by jjfumero**
#### Description

Draft PR - > Proposal to simplify the prebuilt-task API. 

Since we split Task-Graph into different data structures for definition and execution, we do not need all scheduling arguments in the prebuilt task API. This PR proposes a simplification of the Prebuilt Task API for:

1) Create Accessors to define data and its own accessor to the kernel. This is similar to SYCL and oneAPI. 
2) Remove the scheduling parameters from the Prebuilt-Task Graph. This information is stored in a GridScheduler that is passed at a later stage to the execution plan. 

As an example:

```java
       // Define accessors for each parameter
        AccessorParameters accessorParameters = new AccessorParameters(3);
        accessorParameters.set(0, a, Access.READ_ONLY);
        accessorParameters.set(1, b, Access.READ_ONLY);
        accessorParameters.set(2, c, Access.WRITE_ONLY);

        // Define the Task-Graph
        TaskGraph taskGraph = new TaskGraph("s0") //
                .transferToDevice(DataTransferMode.FIRST_EXECUTION, a, b) //
                .prebuiltTask("t0",      //task name
                        "add",              // name of the low-level kernel to invoke
                        FILE_PATH,          // file name
                        accessorParameters) // accessors
                .transferToHost(DataTransferMode.EVERY_EXECUTION, c);

        // When using the prebuilt API, we need to define the WorkerGrid, otherwise it will launch 1 thread
        // on the target device
        WorkerGrid workerGrid = new WorkerGrid1D(numElements);
        GridScheduler gridScheduler = new GridScheduler("s0.t0", workerGrid);

        // Launch the application on the target device
        try (TornadoExecutionPlan executionPlan = new TornadoExecutionPlan(taskGraph.snapshot())) {
            executionPlan.withGridScheduler(gridScheduler) //
                    .withDevice(defaultDevice) //
                    .execute();
        }
```

#### Problem description

n/ a.

#### Backend/s tested

Mark the backends affected by this PR.

- [X] OpenCL
- [X] PTX
- [X] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [X] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If it is applicable, check your changes on FPGAs.

- [ ] Yes
- [X] No

#### How to test the new patch?

```bash
tornado-test --debug -V uk.ac.manchester.tornado.unittests.prebuilt.PrebuiltTest

tornado-test -pk --threadInfo -V --fast uk.ac.manchester.tornado.unittests.atomics.TestAtomics#testAtomic05_precompiled
```

### CODE REVIEW

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/AccessorParameters.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * limitations under the License.
+ *
+ */
+package uk.ac.manchester.tornado.api;
+
+import uk.ac.manchester.tornado.api.common.Access;
+
+/**
+ * Definition of accessor for the Prebuilt-Task API.
+ * It defines for all input parameter to the task, the accessor of each object.*
```
Comment (stratika) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/AccessorParameters.java:24]:
```suggestion
 * It defines the accessor of each object for all input parameter to the task.
```

File: tornado-benchmarks/src/main/java/uk/ac/manchester/tornado/benchmarks/dgemm/DgemmTornado.java [hunk 2]
```diff
@@ -72,39 +69,11 @@ public void setUp() {
...
-                            new int[] { n, n })//
-                    .transferToHost(DataTransferMode.EVERY_EXECUTION, c); //
-            immutableTaskGraph = taskGraph.snapshot();
-            executionPlan = new TornadoExecutionPlan(immutableTaskGraph);
-        }
+        taskGraph.transferToDevice(DataTransferMode.EVERY_EXECUTION, a, b) //
+                .task("dgemm", LinearAlgebraArrays::dgemm, m, n, n, a, b, c) //
+                .transferToHost(DataTransferMode.EVERY_EXECUTION, c);
+        executionPlan = new TornadoExecutionPlan(taskGraph.snapshot());
+        executionPlan.withWarmUp();
```
Comment (stratika) [inline tornado-benchmarks/src/main/java/uk/ac/manchester/tornado/benchmarks/dgemm/DgemmTornado.java:76]: in this case, we removed completely the prebuilt API usage for this benchmark?

File: tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/analyzer/TaskUtils.java [hunk 3]
```diff
@@ -312,15 +312,14 @@ private static DomainTree buildDomainTree(int[] dims) {
     }
 
     public static PrebuiltTask createTask(ScheduleMetaData meta, PrebuiltTaskPackage taskPackage) {
-        DomainTree domain = buildDomainTree(taskPackage.getDimensions());
+        //DomainTree domain = buildDomainTree(taskPackage.getDimensions());
+        DomainTree domain = null;
```
Comment (stratika) [inline tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/analyzer/TaskUtils.java:316]: Remove the domain. I think it is not used any more.

Comment (stratika) [reply tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/analyzer/TaskUtils.java:316]: Do we need the `buildDomainTree` method?

File: tornado-examples/src/main/java/uk/ac/manchester/tornado/examples/arrays/ArrayAddIntPrebuilt.java [hunk 4]
```diff
@@ -62,18 +70,27 @@ public static void main(String[] args) {
...
-        TornadoExecutionPlan executor = new TornadoExecutionPlan(immutableTaskGraph);
-        executor.execute();
+        WorkerGrid workerGrid = new WorkerGrid1D(numElements);
+        GridScheduler gridScheduler = new GridScheduler("s0.t0", workerGrid);
+
+        try (TornadoExecutionPlan executor = new TornadoExecutionPlan(taskGraph.snapshot())) {
+            executor.withDevice(device) //
+                    .withGridScheduler(gridScheduler) //
+                    .execute();
+
```
Comment (stratika) [inline tornado-examples/src/main/java/uk/ac/manchester/tornado/examples/arrays/ArrayAddIntPrebuilt.java:90]: Remove this line.

In this example, I get the message:
```java
a: uk.ac.manchester.tornado.api.types.arrays.IntArray@4f6f416f
b: uk.ac.manchester.tornado.api.types.arrays.IntArray@6d07a63d
c: uk.ac.manchester.tornado.api.types.arrays.IntArray@1bd39d3c
```

Do we need to print this message? or is it a leftover?

**Review by stratika — CHANGES_REQUESTED** [review]

File: tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/analyzer/TaskUtils.java (hunk 3 above)
Comment (jjfumero) [reply tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/analyzer/TaskUtils.java:316]: Yes, we can remove the `domain` at this point.

File: tornado-examples/src/main/java/uk/ac/manchester/tornado/examples/arrays/ArrayAddIntPrebuilt.java (hunk 4 above)
Comment (jjfumero) [reply tornado-examples/src/main/java/uk/ac/manchester/tornado/examples/arrays/ArrayAddIntPrebuilt.java:90]: It was like that but we can remove it.

File: tornado-benchmarks/src/main/java/uk/ac/manchester/tornado/benchmarks/dgemm/DgemmTornado.java (hunk 2 above)
Comment (jjfumero) [reply tornado-benchmarks/src/main/java/uk/ac/manchester/tornado/benchmarks/dgemm/DgemmTornado.java:76]: Yes, that was for an experiment with did for Intel, but do not need to keep this in this version.

**Comment by jjfumero** [comment]
All comments applied.

**Review by stratika — APPROVED** [review]
LGTM

**Review by mairooni — APPROVED** [review]
LGTM

**Review by mikepapadim — APPROVED** [review]

**Pull request merged**